### PR TITLE
migrations: Organize migration files into directories

### PR DESCRIPTION
**Pull request by efritz**
Pulled from #29831. This PR updates migrations to be a pair of up/down files to a rigidly-structured directories where we can add more metadata and migration-specific docs (that are eventually gc'd on squash). This PR also:

- updates the `internal/database/migration/definitions` package to read from directories
- updates `sg migration {add,squash}` to generate the new format
- updates the go-backcompat check to work ensure we can still test on commits with flat-listed definitions (for now)

**Reviewers**: Please review by-commit. Commits that rewrite migrations were done mechanically with the temporary scripts added + removed in this PR.

### CODE REVIEW

**Review by eseliger — APPROVED** [review]
I like this new format!

**Comment by coury-clark** [comment]
Is this an example `internal/database/migration/definition/testdata/missing-downgrade-query/10004/metadata.yaml` of the new file format?

In this I assume `testdata` could be the name of a DB target. `missing-downgrade-query` is the name of a migration, but there are sequential numbered migrations below that?

**Comment by efritz** [comment]
@coury-clark 

The schema's root path is `internal/database/migration/definition/testdata/missing-downgrade-query`, the migration version is `10004`, and the definition of that migration is controlled by the content of `{up,down,metadata.yaml}` (plus anything additional you'd like to co-locate).

So basically instead of just having `{version}_{name}.{up,down}.sql` we have `{version}/{up,down,metadata.sql}`. Versions will still be relatively increasing (probably a timestamp), but we'll be able to break the implicit order dependence by expressing more in the metadata file (which will become messy if we keep it embedded in the up.sql file).

For example, we could solve the issue in https://github.com/sourcegraph/sourcegraph/pull/30306 by also including a `privileged.{up,down}.sql` that can be additionally applied as part of up/down processing (cc @kevinwojo).

**Comment by coury-clark** [comment]
@efritz just curious - why doesn't `version` come before the path? It just seems a bit confusing to me that versions are still increasing (globally? within a db?) but this hierarchy sort of implies multiple versions fall under a single name. I guess the inverse is that multiple names fall under a single version and that doesn't really make much sense either.

Is there a reason not to just embed the version number in the directory name in a similar way we solve for files now, to remove the implicit (but not really) hierarchy?

**Comment by efritz** [comment]
@coury-clark 

The path you selected is in `testdata`, so we're testing different types of broken schemas. There's no "hierarchy" of migrations here (kept the branch name but renamed the PR after realizing the misnomer).

The schemas are still `frontend`, `codeintel`, `insights`.

**Review by coury-clark — APPROVED** [review]

File: dev/sg/internal/migration/util.go [hunk 1]
```diff
@@ -81,3 +81,27 @@ func migrationDirectoryForDatabase(database db.Database) (string, error) {
 
 	return filepath.Join(repoRoot, "migrations", database.Name), nil
 }
+
+// writeMigrationFiles writes the contents of migrationFileTemplate to the given filepaths.
+func writeMigrationFiles(contents map[string]string) (err error) {
+	defer func() {
+		if err != nil {
+			for path := range contents {
+				// undo any changes to the fs on error
```
Comment (tjdevries) [inline dev/sg/internal/migration/util.go:90]: nice

**Comment by tjdevries** [comment]
Makes sense to me :+1: I like that it's in directories now (did not like pairs of files when I worked with this last time)

**Review by camdencheek — APPROVED** [review]

**Pull request merged**